language/move-vm/runtime/src/runtime.rs

```diff
@@ -60,59 +61,157 @@ impl VMRuntime {
         // changing the bytecode format to include an `is_upgradable` flag in the CompiledModule.
-        if data_store.exists_module(&module_id)? {
-            let old_module_ref =
-                self.loader
-                    .load_module_expect_not_missing(&module_id, data_store, log_context)?;
-            let old_module = old_module_ref.module();
-            let old_m = normalized::Module::new(old_module);
-            let new_m = normalized::Module::new(&compiled_module);
-            let compat = Compatibility::check(&old_m, &new_m);
-            if !compat.is_fully_compatible() {
-                return Err(
-                    PartialVMError::new(StatusCode::BACKWARD_INCOMPATIBLE_MODULE_UPDATE)
-                        .finish(Location::Undefined),
-                );
+        for module in &compiled_modules {
+            let module_id = module.self_id();
+            if data_store.exists_module(&module_id)? {
+                let old_module_ref = self.loader.load_module_expect_not_missing(
+                    &module_id,
+                    data_store,
+                    log_context,
+                )?;
+                let old_module = old_module_ref.module();
+                let old_m = normalized::Module::new(old_module);
+                let new_m = normalized::Module::new(&module);
+                let compat = Compatibility::check(&old_m, &new_m);
+                if !compat.is_fully_compatible() {
+                    return Err(PartialVMError::new(
+                        StatusCode::BACKWARD_INCOMPATIBLE_MODULE_UPDATE,
+                    )
+                    .finish(Location::Undefined));
+                }
+            }
+            if !bundle_unverified.insert(module_id) {
+                return Err(PartialVMError::new(StatusCode::DUPLICATE_MODULE_NAME)
+                    .finish(Location::Undefined));
             }
         }
 
-        // perform bytecode and loading verification
-        self.loader
-            .verify_module_for_publication(&compiled_module, data_store, log_context)?;
+        // Perform bytecode and loading verification. Modules must be sorted in topological order.
+        let mut verified_module_ids = vec![];
+        let mut bundle_verified = BTreeMap::new();
+        for module in compiled_modules {
+            let module_id = module.self_id();
+            bundle_unverified.remove(&module_id);
+
+            self.loader.verify_module_for_publication(
+                &module,
+                &bundle_verified,
+                &bundle_unverified,
+                data_store,
+                log_context,
+            )?;
+            bundle_verified.insert(module_id.clone(), module);
+            verified_module_ids.push(module_id);
+        }
 
-        data_store.publish_module(&module_id, module)
+        // NOTE: we want to (informally) argue that all modules pass the linking check before being
```

Comment:
I haven't had time to digest this in detail yet, but there's one test that would greatly increase my confidence: publish a module `A` that contains `friend B` in a vec that does *not* also contain `B`.

Reply:
Tests added in Move CLI testsuite, under dir `multi_module_publish`.

We cannot use `functional-tests` or `e2e-tests` yet because we don't have support for multi-module publishing at the `TransactionPayload` level (`TransactionPayload::Module(<single-module-here>)`). So Move CLI becomes really helpful in this case.